src/main/java/javax/measure/unit/Unit.java

```diff
@@ -201,6 +226,18 @@ public final Dimension getDimension() {
         return dimension;
     }
 
+    /**
+     * The cache will be cleared when this limit is reached, and it will be refilled quickly if necessary.
+     * The intent is to have a fast (and small) cache that speed up the client application.
+     */
+    private static final int CONVERTER_CACHE_SIZE_LIMIT = 100;
+    /**
+     * Cache of unit converters.
+     */
+    private static final ConcurrentHashMap<Unit<?>, ConcurrentHashMap<Unit<?>, UnitConverter>>
```

Comment:
Use a Pair<Unit, Unit> instead of nested maps for faster lookups. I.e. `ConcurrentHashMap<Pair<Unit, Unit>, UnitConverter> CACHE`

Reply:
I think the issue with `Pair` is that it's not part of the core jdk (🤔) - you only find it in the `javafx` package.
Maybe we just do our own stripped-down version of it as suggested in section 5 here: https://www.baeldung.com/java-pairs?

Also, the size of the map will be much larger with a single map. I assume we'd still be faster, but we'll need to hold ~1000s instead of 100.

Reply:
I think Sei tried with Pair and found that it was slower.

Reply:
> I think the issue with Pair is that it's not part of the core jdk 

Guava has one and even if we wrote our own, that's like 20 lines of code, so I'm not concerned :) 

> Also, the size of the map will be much larger with a single map. I assume we'd still be faster, but we'll need to hold ~1000s instead of 100.

Actually, I think the current one is too big. It allows 100 entries on the first level and unlimited entries on the second. I would only cache the 16 or 32 most recently used unit pairs. We don't need to remember every pair that ever happened, we just need to keep the ones in L1 cache that are currently in heavy use by a calculation.

The other problem is that our benchmark is very synthetic. It's constantly hitting this lookup table, so its performance is exaggerated. In practice the lookup table might frequently be pushed out of L1 cache by other things going on in the system and thus have slower performance. Possibly even slower than not having it. This is always a danger when replacing a handful of instructions with a lookup table. You have to be extremely careful with your benchmarks.

> I think Sei tried with Pair and found that it was slower.

I'd love to see those benchmarks and the code.

Reply:
>AbstractMap.SimpleImmutableEntry is in the JDK.  
TIL! this is nice

Reply:
> Use a Pair<Unit, Unit> instead of nested maps for faster lookups

Yes, I tried this initial version because it seemed the most appropriate structure. I had Kotlin class with Pair<Unit, Unit> as key, and the performance measured with the benchmarks was worse than the current structure.  
Not sure if AbstractMap.SimpleImmutableEntry as the key would be faster, but that's something we can measure.

Reply:
Kotlin's Pair does not cache its hash code, that's why I've implemented a custom key class on my branch. I didn't literally mean `Pair`, I just meant a compound key :)